crates/sui-types/src/storage/read_store.rs

```diff
@@ -1,131 +1,645 @@
+    /// "full" checkpoints include: header, contents, transactions, effects
     fn get_full_checkpoint_contents_by_sequence_number(
         &self,
         sequence_number: CheckpointSequenceNumber,
-    ) -> Result<Option<FullCheckpointContents>, Self::Error>;
+    ) -> Result<Option<FullCheckpointContents>>;
 
+    /// Get a "full" checkpoint for purposes of state-sync
+    /// "full" checkpoints include: header, contents, transactions, effects
     fn get_full_checkpoint_contents(
         &self,
         digest: &CheckpointContentsDigest,
-    ) -> Result<Option<FullCheckpointContents>, Self::Error>;
+    ) -> Result<Option<FullCheckpointContents>>;
+
+    // Fetch all checkpint data
+    // TODO fix return type to not be anyhow
+    fn get_checkpoint_data(
+        &self,
+        checkpoint: VerifiedCheckpoint,
+        checkpoint_contents: CheckpointContents,
+    ) -> anyhow::Result<CheckpointData> {
+        use super::ObjectKey;
+        use crate::effects::TransactionEffectsAPI;
+        use crate::full_checkpoint_content::CheckpointTransaction;
+        use std::collections::{HashMap, HashSet};
+
+        let transaction_digests = checkpoint_contents
+            .iter()
+            .map(|execution_digests| execution_digests.transaction)
+            .collect::<Vec<_>>();
+        let transactions = self
+            .multi_get_transactions(&transaction_digests)?
+            .into_iter()
+            .map(|maybe_transaction| {
+                maybe_transaction.ok_or_else(|| anyhow::anyhow!("missing transaction"))
+            })
+            .collect::<anyhow::Result<Vec<_>>>()?;
+
+        let effects = self
+            .multi_get_transaction_effects(&transaction_digests)?
+            .into_iter()
+            .map(|maybe_effects| maybe_effects.ok_or_else(|| anyhow::anyhow!("missing effects")))
+            .collect::<anyhow::Result<Vec<_>>>()?;
+
+        let event_digests = effects
+            .iter()
+            .flat_map(|fx| fx.events_digest().copied())
+            .collect::<Vec<_>>();
+
+        let events = self
+            .multi_get_events(&event_digests)?
+            .into_iter()
+            .map(|maybe_event| maybe_event.ok_or_else(|| anyhow::anyhow!("missing event")))
+            .collect::<anyhow::Result<Vec<_>>>()?;
+
+        let events = event_digests
+            .into_iter()
+            .zip(events)
+            .collect::<HashMap<_, _>>();
+        let mut full_transactions = Vec::with_capacity(transactions.len());
+        for (tx, fx) in transactions.into_iter().zip(effects) {
+            let events = fx.events_digest().map(|event_digest| {
+                events
+                    .get(event_digest)
+                    .cloned()
+                    .expect("event was already checked to be present")
+            });
+            // Note unwrapped_then_deleted contains **updated** versions.
+            let unwrapped_then_deleted_obj_ids = fx
```

Comment:
This isn't really new from this PR, its just moved code.

But i don't remember the answer. I think its because it lets us differentiate before the version pre deleted and the point at which it was deleted